File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    connections = [[] for _ in range(n + 1)]
    for a, b in wires:
        connections[a].append(b)
        connections[b].append(a)
        
    def bfs(start, exclude):
        visited = [False] * (n + 1)
        queue = deque([start])
        visited[start] = True
        count = 1
        
        while queue:
            node = queue.popleft()
            for neighbor in connections[node]:
                if neighbor != exclude and not visited[neighbor]:
                    visited[neighbor] = True
                    queue.append(neighbor)
                    count += 1
        return count
    
    imbalance = float('inf')
    for wire in wires:
        count_a = bfs(wire[0], wire[1])
        count_b = bfs(wire[1], wire[0])
        imbalance = min(imbalance, abs(count_a - count_b))
        if imbalance == 0:
            return 0
            
    return imbalance

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static List<List<Integer>> graph;
    public int solution(int n, int[][] wires) {
        int mini = Integer.MAX_VALUE;
        
        graph = new ArrayList<>();
        for(int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        
        for(int[] wire : wires) {
            graph.get(wire[0]).add(wire[1]);
            graph.get(wire[1]).add(wire[0]);
        }
        
        for(int[] wire : wires) {
            graph.get(wire[0]).remove((Integer) wire[1]);
            graph.get(wire[1]).remove((Integer) wire[0]);
            int diff = Math.abs(checkNum(wire[0], n) - checkNum(wire[1], n));
            mini = Math.min(mini, diff);
            graph.get(wire[0]).add(wire[1]);
            graph.get(wire[1]).add(wire[0]);
        }
        return mini;
    }
    
    public int checkNum(int start, int n) {
        Queue<Integer> q = new LinkedList<>();
        boolean[] visited = new boolean[n + 1];
        int count = 1;
        
        q.offer(start);
        visited[start] = true;
        
        while(!q.isEmpty()) {
            int cur = q.poll();
            for(int next : graph.get(cur)) {
                if(!visited[next]) {
                    visited[next] = true;
                    count++;
                    q.offer(next);
                }
            }
        }
        return count;
    }
}